Followup PR for bad #11624 PR

Quix0r opened the pull request:
This PR fixes introduced by #11624 newly appearing `TypeError` messages:

- fixed TypeError for HTML::toBBCode() invocations with NULL as first (wrong) argument, thanks to @tobias@social.diekershoff.de pointing this out.
- added more checked type-hints as they prevent bad method invocations
- added missing documentation for parameter

Discussion in Friendica forum visible by him: https://social.diekershoff.de/display/f3ad7b1c-9462-ab08-2c0b-330449088287

PS: I know my instance gives "Access forbidden" back.

Quix0r committed:
Fixes and type-hints:
- added more checked type-hints as they prevent bad method invocations
- fixed TypeError for HTML::toBBCode() invocations with NULL as first (wrong)
  argument, thanks to @tobias@social.diekershoff.de pointing this out.

Quix0r committed:
Ops:
- wrong way around (!empty($foo)) is proper
- also needed to be checked on $acitivty['content']

Quix0r commented:
Ops, not complete! Commit is coming up.

Quix0r committed:
Fixes for bad invocations of HTML::toBBCode() (1st parameter is now string)

annando commented on src/Model/APContact.php:
```diff
@@ -241,9 +241,13 @@ public static function getByURL($url, $update = null)
 		}
 
-		$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value'));
+		$aboutHtml = JsonLD::fetchElement($compacted, 'as:summary', '@value');
```
Like in the other case you can make it this way:
```
$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value') ?? '');
```

Quix0r replied:
I currently see no alternative here but it still runs "expensive" code (in `HTML::toBBCode()`) for an empty string which is absolutely useless.

MrPetovan commented on src/Protocol/ActivityPub/Processor.php:
```diff
@@ -615,8 +616,8 @@ private static function processContent($activity, $item)
 			// By default assume "text/html"
-			$item['title'] = HTML::toBBCode($activity['name']);
-			$content = HTML::toBBCode($activity['content']);
+			$item['title'] = (!empty($activity['name']) ? HTML::toBBCode($activity['name']) : '');
+			$content = (!empty($activity['content']) ? HTML::toBBCode($activity['content']) : '');
```
```suggestion
			$item['title'] = HTML::toBBCode($activity['name'] ?? '');
			$content = HTML::toBBCode($activity['content'] ?? '');
```

Quix0r replied:
This suggested change will invoke `HTML::toBBCode()` with an empty `$message` parameter which is completely useless here as to much code runs. I would also suggest to put a check on empty `$message` into `toBBCode()` to avoid executing all that "expensive" code just for an empty string.

MrPetovan replied:
A single check in `HTML::toBBCode` rather than a check for each invocation sounds good to me.

Quix0r replied:
Okay, I added that code already (not yet committed). The changes I made here were caused by `$messages` being `null` from these 2 sources:

1) `APContact.php`, line 214 (old code)
2) `Processor.php`, starting with line 618

Both invoked `HTML::toBBCode()` with `null` for `$message`.

Quix0r replied:
I "resolved" all below as they are basically the same as this one.

MrPetovan replied:
Now that you've added the check for empty string in the `HTML::toBBCode` method, please simplify these conditions with the null-coalescing operator, including both below.

Quix0r replied:
https://www.phptutorial.net/php-tutorial/php-null-coalescing-operator/
I didn't know that `??` is the null-coalescing operator.

Quix0r replied:
Na, I screwed it with rebase on this `develop` branch and forced-push. :-(

MrPetovan commented on src/Protocol/ActivityPub/Processor.php:
```diff
@@ -650,7 +651,7 @@ private static function processContent($activity, $item)
 				$content = self::removeImplicitMentionsFromBody($content, $parent);
 			}
-			$item['content-warning'] = HTML::toBBCode($activity['summary']);
+			$item['content-warning'] = (!empty($activity['summary']) ? HTML::toBBCode($activity['summary']) : '');
```
```suggestion
			$item['content-warning'] = HTML::toBBCode($activity['summary'] ?? '');
```

Quix0r replied:
The same here. It would execute a lot code for an empty string.

MrPetovan commented on src/Protocol/ActivityPub/Processor.php:
```diff
@@ -1224,7 +1225,8 @@ private static function acceptIncomingMessage(array $activity, string $id)
 		$authorid = Contact::getIdForURL($attributed_to);
 
-		$body = HTML::toBBCode(JsonLD::fetchElement($activity['as:object'], 'as:content', '@value'));
+		$value = JsonLD::fetchElement($activity['as:object'], 'as:content', '@value');
+		$body = (!empty($value) ? HTML::toBBCode($value) : '');
```
```suggestion
		$body = HTML::toBBCode(JsonLD::fetchElement($activity['as:object'], 'as:content', '@value') ?? '');
```

Quix0r replied:
The same here again.

MrPetovan replied:
Please reconsider this.
```suggestion
		$body = HTML::toBBCode(JsonLD::fetchElement($activity['as:object'], 'as:content', '@value') ?? '');
```

MrPetovan commented on src/Model/APContact.php:
```diff
@@ -241,9 +241,13 @@ public static function getByURL($url, $update = null)
 		}
 
-		$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value'));
+		$aboutHtml = JsonLD::fetchElement($compacted, 'as:summary', '@value');
+		if ($aboutHtml != "") {
+			$apcontact['about'] = HTML::toBBCode($aboutHtml);
+		}
```
```suggestion
		$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value') ?? '');
```

Quix0r replied:
The same here again.

Quix0r committed:
This will 2 things:
1) The first change prevents "expensive" code in HTML::toBBcode() to be executed
   just for an empty string which makes no sense.
2) The above change was maybe flawed as $apcontact['about'] would have never
   been created, not even empty which could have side effects

Thanks to @annando to make me rethink this part of code.

Quix0r commented:
> which makes no sense.

I mean here executing such an amount of code present in `HTML::toBBCode()` only for an empty string makes no sense.

Quix0r committed:
Added check as suggested by @MrPetovan for empty $message.

annando commented on src/Content/Text/HTML.php:
```diff
@@ -143,6 +143,14 @@ private static function tagToBBCodeSub(DOMDocument $doc, string $tag, array $att
 	 */
 	public static function toBBCode(string $message, string $basepath = ''): string
```
When you make it this way, then even `null` is accepted:
```
public static function toBBCode(string $message = null, string $basepath = ''): string
```

MrPetovan replied:
Technically it's true but I don't think it should be this way. This is a string processing method, not a null processing method.

Quix0r replied:
Yes, it does. On the other hand it makes `$message` optional, too. So even an invocation with no parameters set it is allowed/possible.

Quix0r committed:
Changed to null-coalscing style (??) as sugguested by @MrPetovan

MrPetovan commented:
> Parse error: syntax error, unexpected ':' in ./src/Protocol/ActivityPub/Processor.php on line 619

Quix0r committed:
Ops, syntax errors get unnoticed with a simple editor. :-(

MrPetovan commented on src/Model/APContact.php:
```diff
@@ -241,9 +241,11 @@ public static function getByURL($url, $update = null)
 		}
 
-		$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value'));
+		$aboutHtml = JsonLD::fetchElement($compacted, 'as:summary', '@value');
+		$apcontact['about'] = (!empty($aboutHtml) ? HTML::toBBCode($aboutHtml) : '');
```
```suggestion
		$apcontact['about'] = HTML::toBBCode(JsonLD::fetchElement($compacted, 'as:summary', '@value') ?? '');
```

Quix0r replied:
Done and thank you for your patience. :-)

Quix0r committed:
Also make this null-coalscing

Quix0r committed:
Made also this one sweeter (null-coalscing)

Quix0r commented:
Good work everybody. @MrPetovan @annando :+1:
